File: array-threshold.cpp
//Time - O(nlogn)
//Space - O(1)
class ArrayThreshold {
    public:
    float findArrayThreshold(vector<int> arr, int sum){
        float len = arr.size();
        float threshold = sum/len;
        sort(arr.begin(),arr.end());
        float remaining = 0;
        for(int i = 0;i<len;i++){
            if(threshold>arr[i]){
                remaining = threshold - arr[i];
                threshold = threshold + remaining/(len-i-1);
            }else{
                break;
            }
        }
        
        return threshold;
        
    }
};

int main() {
    std::cout << "Hello World!\n";
    ArrayThreshold t;
    vector<int> nums {2,1,3};
    cout<<t.findArrayThreshold(nums,6)<<endl;
    cout<<t.findArrayThreshold(nums,4)<<endl;
    vector<int> b {2,1,2,6,3,1};
    cout<<t.findArrayThreshold(b,10)<<endl;
}

File: Array Threshold.py
class Solution:
    def arrayThreshold(self, nums, desiredSum):
        #Approach: One-pass
        #Time Complexity: O(n)
        #Space Complexity: O(1)
        
        if sum(nums) < desiredSum:
            return -1
        
        nums.sort()
        n = len(nums)
        
        thresh = desiredSum / n             # best case
        for i in range(len(nums)):
            if nums[i] >= thresh:           # all remaining nums are greater than thresh
                break
            
            diff = thresh - nums[i]
            thresh += diff / (n - i - 1)    # distribute diff among the remaining nums
            
        return thresh
    
#Driver code
sol = Solution()
print(sol.arrayThreshold([2, 1, 5], 6))
print(sol.arrayThreshold([2, 1, 5], 2))
print(sol.arrayThreshold([2, 1, 5], 4))
print(sol.arrayThreshold([2, 1, 5], 1))
print(sol.arrayThreshold([2, 1, 5], 8))
print(sol.arrayThreshold([2, 1, 5], 9))
print('\n')
print(sol.arrayThreshold([5, 10, 4, 3, 9, 1], 9))
print(sol.arrayThreshold([5, 10, 4, 3, 9, 1], 33))
print(sol.arrayThreshold([5, 10, 4, 3, 9, 1], 25))
print(sol.arrayThreshold([5, 10, 4, 3, 9, 1], 16))
print(sol.arrayThreshold([5, 10, 4, 3, 9, 1], 22))
print(sol.arrayThreshold([5, 10, 4, 3, 9, 1], 6))